storage/fanout.go

```diff
@@ -188,142 +208,146 @@ func (f *fanoutAppender) Rollback() (err error) {
 		}
 	}
 
 	if len(filtered) == 0 {
-		return primaryQuerier
-	}
-
-	if primaryQuerier == nil && len(filtered) == 1 {
-		return &chunkQuerierAdapter{filtered[0]}
+		return primary
 	}
 
 	return &chunkQuerierAdapter{&mergeGenericQuerier{
-		mergeFunc:      (&chunkSeriesMergerAdapter{VerticalChunkSeriesMergerFunc: merger}).Merge,
-		primaryQuerier: newGenericQuerierFromChunk(primaryQuerier),
-		queriers:       filtered,
+		merge:          (&chunkSeriesMergerAdapter{VerticalChunkSeriesMergeFunc: merge}).Merge,
+		primary:        newGenericQuerierFromChunk(primary),
+		secondaries:    filtered,
 		failedQueriers: make(map[genericQuerier]struct{}),
 		setQuerierMap:  make(map[genericSeriesSet]genericQuerier),
 	}}
 }
 
 // Select returns a set of series that matches the given label matchers.
-func (q *mergeGenericQuerier) Select(sortSeries bool, hints *SelectHints, matchers ...*labels.Matcher) (genericSeriesSet, Warnings, error) {
-	if len(q.queriers) == 1 {
-		return q.queriers[0].Select(sortSeries, hints, matchers...)
-	}
-
+func (q *mergeGenericQuerier) Select(_ bool, hints *SelectHints, matchers ...*labels.Matcher) (genericSeriesSet, Warnings, error) {
 	var (
-		seriesSets = make([]genericSeriesSet, 0, len(q.queriers))
+		seriesSets = make([]genericSeriesSet, 0, len(q.secondaries)+1)
+		queries    = append([]genericQuerier{q.primary}, q.secondaries...)
 		warnings   Warnings
 		priErr     error
+		wg         sync.WaitGroup
 	)
+
 	type queryResult struct {
-		qr          genericQuerier
-		set         genericSeriesSet
-		wrn         Warnings
-		selectError error
-	}
-	queryResultChan := make(chan *queryResult)
-
-	for _, querier := range q.queriers {
-		go func(qr genericQuerier) {
-			// We need to sort for NewMergeSeriesSet to work.
-			set, wrn, err := qr.Select(true, hints, matchers...)
-			queryResultChan <- &queryResult{qr: qr, set: set, wrn: wrn, selectError: err}
-		}(querier)
-	}
-	for i := 0; i < len(q.queriers); i++ {
-		qryResult := <-queryResultChan
-		q.setQuerierMap[qryResult.set] = qryResult.qr
-		if qryResult.wrn != nil {
-			warnings = append(warnings, qryResult.wrn...)
-		}
-		if qryResult.selectError != nil {
-			q.failedQueriers[qryResult.qr] = struct{}{}
-			// If the error source isn't the primary querier, return the error as a warning and continue.
-			if !reflect.DeepEqual(qryResult.qr, q.primaryQuerier) {
-				warnings = append(warnings, qryResult.selectError)
-			} else {
-				priErr = qryResult.selectError
+		q   genericQuerier
+		set genericSeriesSet
+		wrn Warnings
+		err error
+	}
+	resultChan := make(chan *queryResult)
+
+	for _, q := range queries {
+		wg.Add(1)
+		go func(q genericQuerier) {
+			// We need to sort for newGenericMergeSeriesSet to work.
+			set, wrn, err := q.Select(true, hints, matchers...)
+			resultChan <- &queryResult{q: q, set: set, wrn: wrn, err: err}
+			wg.Done()
+		}(q)
+	}
+	go func() {
+		wg.Wait()
+		close(resultChan)
+	}()
+
+	for r := range resultChan {
```

Comment:
Few questions here:
1. Don't we need this to be in time-order as we use this with chained merge too?
2. (unrelated to this line) Do we attempt vertical merge for the queries which do not have overlapping data too?

Reply:
1. We don't.  `ChainingSeriesMerge` handles this just fine. But good point to add note to `mergeFunc` that it has to handle underodered time wise series (: Thanks! 
2. No, see:

```
func (c *genericMergeSeriesSet) At() Labels {
	if len(c.currentSets) == 1 {
		return c.currentSets[0].At()
	}
	series := make([]Labels, 0, len(c.currentSets))
	for _, seriesSet := range c.currentSets {
		series = append(series, seriesSet.At())
	}
	return c.mergeFunc(series...)
}

```